### PR TITLE
[DNM]Enhanced Connection Establishment -- v2

### PR DESCRIPTION
**Pull request by changchengx**
Signed-off-by: Changcheng Liu <jerrliu@nvidia.com>

## What
Support ECE(enhanced connection establishment) in UCX.

## Why ?
ECE could be used to negotiate 1) packet loss 2) congestion control strategy etc. between source and target.
It's better to make use of this feature in UCX.
Issue: 2332889 & 2684830

## How ?
Design document
Public: [UCX ECE development](https://docs.google.com/document/d/10VzrdmuWzc4VDbL3cTeO1aBRNWVnZ8rQ6-srmfO4E6o/edit?usp=sharing)
Internal wiki: [UCX ECE development](https://wikinox.mellanox.com/display/SW/UCX+ECE+Development)
Internal sharepoint: ```NBU-UCX --> Documents --> Arch documents --> UCX ECE Development.docx```

### CODE REVIEW

**Comment by leibin2014** [comment]
Please move the unnecessary optimizations/changes to another seperate PR, just keep the ECE related changes.

File: src/uct/ib/rc/accel/rc_mlx5.h [hunk 1]
```diff
@@ -177,7 +178,8 @@ ucs_status_t uct_rc_mlx5_ep_tag_rndv_request(uct_ep_h tl_ep, uct_tag_t tag,
                                              unsigned header_length,
                                              unsigned flags);
 
-ucs_status_t uct_rc_mlx5_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *addr);
+ucs_status_t uct_rc_mlx5_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *addr,
+                                        uint32_t *ece_val);
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5.h:182]: rename all ece_val to ece

File: src/uct/ib/rc/accel/rc_mlx5_devx.c [hunk 2]
```diff
@@ -412,13 +412,22 @@ uct_rc_mlx5_iface_common_devx_connect_qp(uct_rc_mlx5_iface_common_t *iface,
...
+        UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, ece, qp->remote_ece.val);
+    }
     UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, opt_param_mask, opt_param_mask);
 
     status = uct_ib_mlx5_devx_modify_qp(qp, in_2rtr, sizeof(in_2rtr),
                                         out_2rtr, sizeof(out_2rtr));
     if (status != UCS_OK) {
         return status;
     }
+    if (dev->flags & UCT_IB_DEVICE_FLAG_ECE) {
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_devx.c:427]: One space line before

File: src/uct/ib/rc/accel/rc_mlx5_devx.c [hunk 3]
```diff
@@ -412,13 +412,22 @@ uct_rc_mlx5_iface_common_devx_connect_qp(uct_rc_mlx5_iface_common_t *iface,
     UCT_IB_MLX5DV_SET(qpc, qpc, rae, true);
     UCT_IB_MLX5DV_SET(qpc, qpc, min_rnr_nak, iface->super.config.min_rnr_timer);
 
+    if ((dev->flags & UCT_IB_DEVICE_FLAG_ECE) &&
+        iface->super.super.config.ece_cfg.ece_enable) {
+        ucs_debug("init2rtr_qp_in with ece 0x%x", qp->remote_ece.val);
+        UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, ece, qp->remote_ece.val);
+    }
     UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, opt_param_mask, opt_param_mask);
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_devx.c:420]: One space line before

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 4]
```diff
@@ -939,33 +961,32 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
...
+         * such a QP to be initialized with zero send queue length. */
+        memset(&attr, 0, sizeof(attr));
+        uct_rc_mlx5_iface_fill_attr(iface, &attr, 0, &iface->rx.srq);
+        uct_ib_exp_qp_fill_attr(&iface->super.super, &attr.super);
+        status = uct_rc_mlx5_iface_create_qp(iface, &self->tm_qp, NULL, &attr);
         if (status != UCS_OK) {
-            goto err;
+            goto err_destroy_txqp;
         }
+        self->super.local_ece.val = ece_int(self->super.local_ece.val,
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c:977]: One space line before

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 5]
```diff
@@ -350,6 +352,18 @@ ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
...
+        iface->super.super.config.ece_cfg.ece_enable &&
+        (0 == ibv_query_ece(qp->verbs.qp, &ece))) {
+        qp->local_ece.val = ece.options;
+    } else {
+#endif
+        qp->local_ece.val = 0;
+#if HAVE_RDMACM_ECE
+    }
+#endif
+
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:366]: #if HAVE_RDMACM_ECE
    if ((dev->flags & UCT_IB_DEVICE_FLAG_ECE) &&
        iface->super.super.config.ece_cfg.ece_enable &&
        (0 == ibv_query_ece(qp->verbs.qp, &ece))) {
        qp->local_ece.val = ece.options;
    } else {
        qp->local_ece.val = 0;
    }
#else
    qp->local_ece.val = 0;
#endif

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 6]
```diff
@@ -285,12 +286,13 @@ ucs_status_t uct_rc_mlx5_iface_create_qp(uct_rc_mlx5_iface_common_t *iface,
                                          uct_ib_mlx5_txwq_t *txwq,
                                          uct_ib_mlx5_qp_attr_t *attr)
 {
-    uct_ib_iface_t *ib_iface           = &iface->super.super;
+    uct_ib_iface_t *ib_iface          = &iface->super.super;
+    uct_ib_mlx5_md_t *md              = ucs_derived_of(ib_iface->super.md,
+                                                       uct_ib_mlx5_md_t);
+    uct_ib_device_t UCS_V_UNUSED *dev = &md->super.dev;
+    uct_ibv_ece_t UCS_V_UNUSED   ece;
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:293]: uct_ibv_ece_t UCS_V_UNUSED ece;

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 7]
```diff
@@ -974,13 +995,11 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_ep_t, const uct_ep_params_t *params)
...
-                                         IBV_EVENT_QP_LAST_WQE_REACHED,
-                                         self->tx.wq.super.qp_num);
-    uct_rc_iface_remove_qp(&iface->super, self->tx.wq.super.qp_num);
-
-err:
+err_destroy_tmqp:
+    if (UCT_RC_MLX5_TM_ENABLED(iface)) {
+        uct_ib_mlx5_destroy_qp(md, &self->tm_qp);
+    }
+err_destroy_txqp:
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c:1002]: One space line before

File: src/uct/ib/base/ib_verbs.h [hunk 8]
```diff
@@ -19,6 +19,12 @@
 #include <infiniband/verbs_exp.h>
 #endif
 
+#if HAVE_RDMACM_ECE
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_verbs.h:22]: ```ibv_ece``` struct exist only under the condition that ```HAVE_RDMACM_ECE``` is defined

File: src/uct/ib/base/ece.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+#include <ucs/config/types.h>
+#include <ucs/debug/assert.h>
+
+enum  {
+    MLX5_ECE_DISABLED  = 0,
+    MLX5_ECE_VER_1     = 1,
+    MLX5_ECE_VER_SR    = 2,  //selective repeat
+    MLX5_ECE_VER_MAX   = MLX5_ECE_VER_SR,
+};
+
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h:22]: Done

File: src/uct/ib/base/ib_device.h [hunk 10]
```diff
@@ -260,6 +262,12 @@ extern const double uct_ib_qp_rnr_time_ms[];
 ucs_status_t uct_ib_device_port_check(uct_ib_device_t *dev, uint8_t port_num,
                                       unsigned flags);
 
+/**
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_device.h:266]: Done

File: src/uct/ib/rc/accel/rc_mlx5_devx.c (hunk 3 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_devx.c:421]: Done

File: src/uct/ib/rc/accel/rc_mlx5_devx.c (hunk 2 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_devx.c:429]: Done

File: src/uct/ib/rc/accel/rc_mlx5_ep.c (hunk 4 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c:978]: Done

File: src/uct/ib/rc/accel/rc_mlx5_ep.c (hunk 7 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c:1005]: Done

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 5 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c:366]: Done

File: src/uct/base/uct_iface.c [hunk 11]
```diff
@@ -608,15 +608,17 @@ void uct_ep_destroy(uct_ep_h ep)
...
+                                uint32_t *ece_val)
 {
-    return ep->iface->ops.ep_get_address(ep, addr);
+    return ep->iface->ops.ep_get_address(ep, addr, ece_val);
 }
 
 ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr,
-                                  const uct_ep_addr_t *ep_addr)
+                                  const uct_ep_addr_t *ep_addr,
+                                  const uint32_t *ece_val)
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/base/uct_iface.c]: Done

File: src/uct/ib/base/ece.h [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+#include <stdint.h>
+#include <infiniband/verbs.h>
+#include <ucs/config/types.h>
+#include <ucs/debug/assert.h>
+
+enum  {
+    MLX5_ECE_DISABLED  = 0,
+    MLX5_ECE_VER_1     = 1,
+    MLX5_ECE_VER_SR    = 2,  //selective repeat
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h]: Done

File: src/uct/ib/base/ece.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h:22]: Done

File: src/uct/ib/base/ib_device.c [hunk 13]
```diff
@@ -815,6 +815,45 @@ ucs_status_t uct_ib_device_port_check(uct_ib_device_t *dev, uint8_t port_num,
...
 }
 
+void uct_ib_device_ece_check(uct_ib_device_t *dev,
+                             struct ibv_context *ctx, struct ibv_pd *pd)
+{
+#if HAVE_RDMACM_ECE
+    struct ibv_qp *dummy_qp;
+    struct ibv_cq *dummy_cq;
+    struct ibv_qp_init_attr dummy_qp_init_attr;
+    struct ibv_ece dummy_ece = {};
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_device.c]: Done

File: src/uct/ib/base/ib_device.h [hunk 14]
```diff
@@ -87,6 +88,7 @@ enum {
     UCT_IB_DEVICE_FLAG_DC       = UCT_IB_DEVICE_FLAG_DC_V1 |
                                   UCT_IB_DEVICE_FLAG_DC_V2, /* Device supports DC */
     UCT_IB_DEVICE_FLAG_ODP_IMPLICIT = UCS_BIT(9),
+    UCT_IB_DEVICE_FLAG_ECE      = UCS_BIT(10),       /* Device supports ECE */
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_device.h]: Done

File: src/uct/ib/rc/accel/rc_mlx5_devx.c (hunk 3 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_devx.c:421]: Done

File: src/uct/ib/rc/accel/rc_mlx5_ep.c (hunk 4 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c:978]: Done

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 6 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: Done

File: src/uct/ib/rc/accel/rc_mlx5_ep.c (hunk 7 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_ep.c:1005]: Done

File: src/uct/ib/dc/dc_mlx5.c [hunk 15]
```diff
@@ -820,35 +986,69 @@ void uct_dc_mlx5_iface_dcis_destroy(uct_dc_mlx5_iface_t *iface, int max)
...
+
 ucs_status_t
 uct_dc_mlx5_iface_create_dcis(uct_dc_mlx5_iface_t *iface,
                               const uct_dc_mlx5_iface_config_t *config)
 {
-    uint8_t num_paths = iface->super.super.super.num_paths;
-    uct_dc_mlx5_dci_pool_t *dci_pool;
-    int i, pool_index, dci_index;
+    int full_handshake =
+        uct_dc_mlx5_force_full_handshake(iface, config->dci_full_handshake);
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/dc/dc_mlx5.c]: Done

File: src/uct/ib/rc/accel/rc_mlx5.h (hunk 1 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5.h]: Done

File: src/uct/ib/base/ib_verbs.h [hunk 16]
```diff
@@ -19,6 +19,12 @@
...
+typedef struct ibv_ece uct_ibv_ece_t;
+#else
+typedef struct {void *placeholder;} uct_ibv_ece_t;
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_verbs.h]: I've changed it to be in multi-lines.

File: src/uct/ib/base/ece.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+enum  {
+    MLX5_ECE_DISABLED  = 0,
+    MLX5_ECE_VER_1     = 1,
+    MLX5_ECE_VER_SR    = 2,  /* selective repeat */
+    MLX5_ECE_VER_MAX   = MLX5_ECE_VER_SR,
+};
+
+
+typedef struct mlx5_ece_cfg {
+    ucs_on_off_auto_value_t ece_enable;
```
Comment (leibin2014) [inline src/uct/ib/base/ece.h:25]: rename to enable

File: src/uct/ib/base/ece.h [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    ucs_on_off_auto_value_t ece_sr;
```
Comment (leibin2014) [inline src/uct/ib/base/ece.h:26]: rename to sr

File: src/uct/ib/base/ece.h [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        uint32_t rsv : 27;
+        uint32_t ver : 4;
+    } field;
+
+    uint32_t val;
+};
+
+
+typedef struct mlx5_ece {
+    uint32_t ece_enable;
```
Comment (leibin2014) [inline src/uct/ib/base/ece.h:42]: Why the name is exactly  the same as line25? Are they the same thing?

File: src/uct/ib/base/ece.h [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+    union ece_t ece;
+} mlx5_ece_t;
+
+
+/**
+ * ece intersection
+ */
+static UCS_F_ALWAYS_INLINE
+uint32_t ece_int(uint32_t ece_val0, uint32_t ece_val1)
```
Comment (leibin2014) [inline src/uct/ib/base/ece.h:52]: rename to val0 and val1

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 21]
```diff
@@ -651,6 +665,125 @@ int uct_rc_mlx5_iface_is_reachable(const uct_iface_h tl_iface,
...
+                             ib_iface->config.ece_cfg.ece.val,
+                             ep.tm_qp.local_ece.val);
+
+                     if (!(ib_iface->config.ece_cfg.ece.val & 0x1)) {
+                         ib_iface->config.ece_cfg.ece_enable = 0;
+                         ib_iface->config.ece_cfg.ece.val    = 0;
+                     } else {
+                         ib_iface->config.ece_cfg.ece_enable = 1;
+                     }
+                    uct_ib_mlx5_destroy_qp(md, &ep.tm_qp);
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:748]: Add blank line before

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 22]
```diff
@@ -837,6 +970,9 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_t,
         init_attr.flags  |= UCT_IB_TM_SUPPORTED;
     }
 
+    self->super.super.super.config.ece_cfg.ece_enable = 0;
+    self->super.super.super.config.ece_cfg.ece.val = 0;
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:974]: Make = aligned.

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 23]
```diff
@@ -852,7 +988,9 @@ UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_t,
         return status;
     }
 
-    return UCS_OK;
+    status = uct_rc_mlx5_init_ece(&self->super, &config->super.super);
+
+    return status;
```
Comment (leibin2014) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:993]: remove blank line before

File: src/uct/ib/rc/verbs/rc_verbs_iface.c [hunk 24]
```diff
@@ -268,6 +268,8 @@ static UCS_CLASS_INIT_FUNC(uct_rc_verbs_iface_t, uct_md_h tl_md,
     uct_ib_iface_config_t *ib_config    = &config->super.super.super;
     uct_ib_iface_init_attr_t init_attr  = {};
     uct_ib_qp_attr_t attr               = {};
+    uct_ibv_ece_t UCS_V_UNUSED ece      = {};
+    mlx5_ece_cfg_t *conn_ece = &config->super.super.conn_ece;
```
Comment (leibin2014) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c:272]: Make = aligned

File: src/uct/ib/base/ece.h (hunk 19 above)
Comment (changchengx) [reply src/uct/ib/base/ece.h:42]: ```mlx5_ece::ece_enable``` is the combination of ```user config option, mlx5_ece_cfg::ece_enable``` and ```device ece cap```.

Comment (leibin2014) [reply src/uct/ib/base/ece.h:42]: > 
@changchengx Then need different names for them.

File: src/uct/ib/base/ib_verbs.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ib_verbs.h]: Move the alias of struct type after the included header files.

File: src/ucp/wireup/wireup_cm.c [hunk 25]
```diff
@@ -795,6 +810,11 @@ static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
     memcpy(progress_arg->sa_data, remote_data->conn_priv_data,
            remote_data->conn_priv_data_length);
 
+    if (remote_data->field_mask & UCT_CM_REMOTE_DATA_FIELD_OOB_ECE) {
+        ucp_ep->remote_ece = remote_data->ece;
+    } else {
+        ucp_ep->remote_ece = 0;
+    }
     uct_worker_progress_register_safe(worker->uct,
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/ucp/wireup/wireup_cm.c:819]: Done

File: src/uct/ib/base/ece.h (hunk 17 above)
Comment (changchengx) [reply src/uct/ib/base/ece.h]: Done

File: src/uct/ib/base/ece.h (hunk 18 above)
Comment (changchengx) [reply src/uct/ib/base/ece.h]: Done

File: src/uct/ib/base/ece.h (hunk 20 above)
Comment (changchengx) [reply src/uct/ib/base/ece.h]: Done

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 21 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c:747]: Done

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 22 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: Done

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 23 above)
Comment (changchengx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c:991]: Done

File: src/uct/ib/rc/verbs/rc_verbs_iface.c (hunk 24 above)
Comment (changchengx) [reply src/uct/ib/rc/verbs/rc_verbs_iface.c]: Done

File: src/ucp/core/ucp_ep.h [hunk 26]
```diff
@@ -81,6 +81,8 @@ enum {
     UCP_EP_FLAG_INDIRECT_ID            = UCS_BIT(14),/* protocols on this endpoint will send
                                                         indirect endpoint id instead of pointer,
                                                         can be replaced with looking at local ID */
+    UCP_EP_FLAG_OOB_ECE                = UCS_BIT(15),/* protocols on this endpoint will send
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/ucp/core/ucp_ep.h]: Remove ```OOB```. The flag here means that whether ucp_ep needs to get ECE from uct_eps and ucp_ep has ece value.

File: src/ucp/wireup/wireup_cm.c [hunk 27]
```diff
@@ -468,11 +471,16 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
...
+    if (ep->local_ece == 0) {
+        ep->flags &= ~UCP_EP_FLAG_OOB_ECE;
+    }
+
     /* Replay pending requests from the tmp_pending_queue */
     ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
 
     can_fallback = (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) ||
                    (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE);
+    ep->flags   |= UCP_EP_FLAG_OOB_ECE;
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/ucp/wireup/wireup_cm.c]: remove ```line 474```. Pass local verification.

File: src/ucp/wireup/wireup_cm.c [hunk 28]
```diff
@@ -795,6 +810,11 @@ static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
     memcpy(progress_arg->sa_data, remote_data->conn_priv_data,
            remote_data->conn_priv_data_length);
 
+    if (remote_data->field_mask & UCT_CM_REMOTE_DATA_FIELD_OOB_ECE) {
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/ucp/wireup/wireup_cm.c]: Remove ```OOB```. ```UCT_CM_REMOTE_DATA_FIELD_OOB_ECE``` means that whether the ```remote_data``` has ece info.

File: src/uct/api/uct_def.h [hunk 29]
```diff
@@ -304,6 +307,9 @@ typedef struct uct_cm_remote_data {
      * Length of the received data from the peer.
      */
     size_t                  conn_priv_data_length;
+
+    /** OOB ECE of the remote peer. */
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/api/uct_def.h]: remove ```OOB```

File: src/uct/ib/base/ece.h (hunk 19 above)
Comment (changchengx) [reply src/uct/ib/base/ece.h]: refine to be ```uint8_t enable``` and add comments about ```struct mxl5_ece``` and ```struct mlx5_ece_cfg```

File: src/uct/ib/dc/dc_mlx5.c [hunk 30]
```diff
@@ -414,21 +444,29 @@ uct_dc_mlx5_force_full_handshake(uct_dc_mlx5_iface_t *iface, int full_handshake_
 ucs_status_t uct_dc_mlx5_iface_dci_connect(uct_dc_mlx5_iface_t *iface,
                                            uct_dc_dci_t *dci)
 {
-    uct_ib_mlx5_md_t *md = ucs_derived_of(iface->super.super.super.super.md,
-                                          uct_ib_mlx5_md_t);
+    uct_ib_iface_t *ib_iface = &iface->super.super.super;
+    uct_ib_mlx5_md_t *md     = ucs_derived_of(ib_iface->super.md,
+                                              uct_ib_mlx5_md_t);
+    uct_ib_device_t UCS_V_UNUSED *dev = uct_ib_iface_device(ib_iface);
```
Comment (leibin2014) [inline src/uct/ib/dc/dc_mlx5.c:450]: Make = align

File: src/uct/ib/dc/dc_mlx5.c [hunk 31]
```diff
@@ -414,21 +444,29 @@ uct_dc_mlx5_force_full_handshake(uct_dc_mlx5_iface_t *iface, int full_handshake_
...
     struct ibv_qp_attr attr;
     long attr_mask;
 
     if (md->flags & UCT_IB_MLX5_MD_FLAG_DEVX) {
+        if (dci->local_ece.val == 0) {
+            dci->txwq.super.remote_ece.val = 0;
+        } else {
+            dci->txwq.super.remote_ece = dci->txwq.super.local_ece;
+        }
         return uct_dc_mlx5_iface_devx_dci_connect(iface, &dci->txwq.super,
```
Comment (leibin2014) [inline src/uct/ib/dc/dc_mlx5.c:461]: Space line before.

File: src/uct/ib/dc/dc_mlx5.c [hunk 32]
```diff
@@ -994,48 +1329,81 @@ ucs_status_t uct_dc_mlx5_iface_flush(uct_iface_h tl_iface, unsigned flags, uct_c
...
 
+err:
+    ep_idx = ep_idx - 1;
 err_cleanup:
-    UCS_CLASS_CLEANUP(uct_base_ep_t, &ep->super);
+    while (idx <= ep_idx) {
+        UCS_CLASS_CLEANUP(uct_base_ep_t, &fc_eps[idx].super);
+        idx = idx + 1;
+    }
+    return status;
```
Comment (leibin2014) [inline src/uct/ib/dc/dc_mlx5.c:1373]: space line before

File: src/uct/ib/dc/dc_mlx5.c [hunk 33]
```diff
@@ -1487,7 +1878,10 @@ uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
     ep = ucs_derived_of(op->ep, uct_dc_mlx5_ep_t);
 
     if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        ucs_assert(ep != iface->tx.fc_ep);
+        ucs_assert(ep != &iface->tx.fc_ep[0]);
+        if (iface->gp != 1) {
+            ucs_assert(ep != &iface->tx.fc_ep[1]);
+        }
         uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
```
Comment (leibin2014) [inline src/uct/ib/dc/dc_mlx5.c:1885]: space line before

File: src/uct/ib/base/ece.h [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+    ucs_on_off_auto_value_t enable;
+    ucs_on_off_auto_value_t sr;
+} mlx5_ece_cfg_t;
+
+
+union ece_t {
+    struct {
+        uint32_t sr  : 1;
+        uint32_t rsv : 27;
+        uint32_t ver : 4;
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h:35]: I'll try to add version3/CC support here before getting Yossi's reply.

File: src/ucp/wireup/address.c [hunk 35]
```diff
@@ -1288,7 +1289,7 @@ ucp_address_do_pack(ucp_worker_h worker, ucp_ep_h ep, void *buffer, size_t size,
                                                      0);
 
                     /* pack ep address */
-                    status = uct_ep_get_address(ep->uct_eps[lane], ptr);
+                    status = uct_ep_get_address(ep->uct_eps[lane], ptr, &ece);
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/ucp/wireup/address.c:1292]: Done

File: src/uct/ib/base/ece.h (hunk 34 above)
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h:38]: Done

File: src/uct/ib/dc/dc_mlx5.c (hunk 30 above)
Comment (changchengx) [reply src/uct/ib/dc/dc_mlx5.c:450]: Done

File: src/uct/ib/dc/dc_mlx5.c (hunk 31 above)
Comment (changchengx) [reply src/uct/ib/dc/dc_mlx5.c:462]: Done

File: src/uct/ib/dc/dc_mlx5.c (hunk 32 above)
Comment (changchengx) [reply src/uct/ib/dc/dc_mlx5.c:1389]: Done

File: src/uct/ib/dc/dc_mlx5.c (hunk 33 above)
Comment (changchengx) [reply src/uct/ib/dc/dc_mlx5.c:1902]: Done

File: src/uct/ib/base/ece.h [hunk 36]
```diff
@@ -0,0 +1,77 @@
...
+#include <stdint.h>
+#include <infiniband/verbs.h>
+#include <ucs/config/types.h>
+#include <ucs/debug/assert.h>
+
+
+enum  {
+    MLX5_ECE_DISABLED  = 0,
+    MLX5_ECE_VER_1     = 1,
+    MLX5_ECE_VER_SR    = 2,  /* selective repeat */
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h]: Done

File: src/uct/ib/base/ece.h [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+
+    union ece_t ece;
+} mlx5_ece_t;
+
+
+/**
+ * ece intersection
+ */
+static UCS_F_ALWAYS_INLINE
+uint32_t ece_int(uint32_t val0, uint32_t val1)
```
[the remark this answers is not in the document]
Comment (changchengx) [reply src/uct/ib/base/ece.h]: change to be ```ece_intersect``` here

**Comment by changchengx** [comment]
@yosefe 
1. UCX ECE design document is in wiki page: [UCX+ECE+Development](https://wikinox.mellanox.com/display/SW/UCX+ECE+Development) or sharepoint: ```NBU-UCX --> Documents --> Arch documents --> UCX ECE Development.docx```
2. UCX ECE code has been implemented in this PR by following the above design document

**Comment by changchengx** [comment]
Close this PR, this is the 2nd version of ECE design which has been rejected internally. The corresponding test case PR is [PR#7980](https://github.com/openucx/ucx/pull/7980) which is closed now. 
The [3rd version of ECE design](https://github.com/openucx/ucx/pull/8096) is approved and merged.

**Pull request closed without merging**